### PR TITLE
Connect napari events first to EventEmitter

### LINKED ISSUE #4421: Move layer triggers weird bug in viewer
## 🐛 Bug

When I want to move a layer, a strange bug is happening in the viewer. I don't know if it's related to my environment, if I am doing something wrong. But here is what happens: when I move a layer in the viewer, I get this error:

```
> ---------------------------------------------------------------------------
> AttributeError                            Traceback (most recent call last)
> /usr/local/lib/python3.10/site-packages/napari/_qt/containers/_layer_delegate.py in paint(self=<napari._qt.containers._layer_delegate.LayerDelegate object>, painter=<PyQt5.QtGui.QPainter object>, option=<PyQt5.QtWidgets.QStyleOptionViewItem object>, index=<PyQt5.QtCore.QModelIndex object>)
>      84         # update the icon based on layer type
>      85 
> ---> 86         self.get_layer_icon(option, index)
>         self.get_layer_icon = <bound method LayerDelegate.get_layer_icon of <napari._qt.containers._layer_delegate.LayerDelegate object at 0x7f193fc9b7f0>>
>         option = <PyQt5.QtWidgets.QStyleOptionViewItem object at 0x7f1901960040>
>         index = <PyQt5.QtCore.QModelIndex object at 0x7f18ffd2f450>
>      87         # paint the standard itemView (includes name, icon, and vis. checkbox)
>      88         super().paint(painter, option, index)
> 
> /usr/local/lib/python3.10/site-packages/napari/_qt/containers/_layer_delegate.py in get_layer_icon(self=<napari._qt.containers._layer_delegate.LayerDelegate object>, option=<PyQt5.QtWidgets.QStyleOptionViewItem object>, index=<PyQt5.QtCore.QModelIndex object>)
>      97             icon_name = 'folder-open' if expanded else 'folder'
>      98         else:
> ---> 99             icon_name = f'new_{layer._type_string}'
>         icon_name = undefined
>     100 
>     101         try:
> 
> AttributeError: 'NoneType' object has no attribute '_type_string'
```
In addition, there is the same error as #4420: the layer remains in the foreground, even if it has been moved from the last layer to a lower layer.

## To Reproduce

1. Move layer from on old index to a new index
`viewer.layers.move(old_index, new_index)`

## Expected behavior
This error  `AttributeError: 'NoneType' object has no attribute '_type_string' ` should not be raised...
And

## Environment
napari: 0.4.15
Platform: Linux-5.17.4-200.fc35.x86_64-x86_64-with-glibc2.34
System: Fedora Linux 35 (Workstation Edition)
Python: 3.10.4 (main, Mar 25 2022, 00:00:00) [GCC 11.2.1 20220127 (Red Hat 11.2.1-9)]
Qt: 5.15.2
PyQt5: 5.15.6
NumPy: 1.21.5
SciPy: 1.8.0
Dask: 2022.04.0
VisPy: 0.9.6

OpenGL:
- GL version: 4.6.0 NVIDIA 510.47.03
- MAX_TEXTURE_SIZE: 16384

Screens:
- screen 1: resolution 1920x1200, scale 1.0

## Additional context
This bug appears in the context of the development of a new plugin (this logic is triggered in a widget). I have another bug here #4420 which might be related.

### LINKED ISSUE #4425: Modification of evented containers in function connected to its events is not safe
## 🐛 Bug

When modify Evented container inside function connected to to this container new events are processed immediately, and then are processed events from original event. 

This may end with inconsistent state. Especially dangerous where more than one such callback is connected to given container (for example layer list) from different plugins. 

## To Reproduce

See issue #4421

## Expected behavior

Container modification in event callback should be delayed, or at least events created by this modification should be processed after callbacks from original event was called.

### PR DESCRIPTION
**Pull request by Czaki**
# Description

This PR ensures that napari callbacks events are triggered first before callbacks from external packages. Thanks to this reorder developers could faster get failing tests connected with real code bugs. Also using `--maxfail=n` have more sense in such a configuration. 

## Type of change

- [x] Bug-fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update

# References

close #4425 
close #4421

# How has this been tested?

- [ ] example: the test suite for my feature covers cases x, y, and z
- [ ] example: all tests pass with my change

## Final checklist:
- [ ] My PR is the minimum possible work for the desired functionality
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] If I included new strings, I have used `trans.` to make them localizable.
      For more information see our [translations guide](https://napari.org/docs/dev/guides/translations.html).

### CODE REVIEW

File: napari/utils/_tests/test_notification_manager.py [hunk 1]
```diff
@@ -75,8 +76,12 @@ def test_notification_manager_no_gui(monkeypatch):
 
         # pytest intercepts the error, so we can manually call sys.excepthook
         assert sys.excepthook == notification_manager.receive_error
-        sys.excepthook(*sys.exc_info())
+        try:
+            raise ValueError("a")
+        except ValueError:
+            sys.excepthook(*sys.exc_info())
```
Comment (Czaki) [inline napari/utils/_tests/test_notification_manager.py:80]: `sys.exc_info()` output was `(None, None, None)` that crash printing exception info and preventing to run all callbacks.

File: napari/utils/_tests/test_notification_manager.py [hunk 2]
```diff
@@ -138,10 +143,13 @@ def _raise():
...
-            threading.excepthook(sys.exc_info())
-        else:
-            sys.excepthook(*sys.exc_info())
+        try:
+            raise ValueError("a")
+        except ValueError:
+            if PY38_OR_HIGHER:
+                threading.excepthook(sys.exc_info())
+            else:
+                sys.excepthook(*sys.exc_info())
```
Comment (jni) [inline napari/utils/_tests/test_notification_manager.py]: We no longer support 3.7, so I guess this can be dropped?

File: napari/utils/events/event.py [hunk 3]
```diff
@@ -369,6 +369,20 @@ def source(self) -> Any:
...
+                callback = callback.func
+            if not isinstance(callback, tuple):
+                return callback.__module__.startswith('napari.')
+            obj = callback[0]()
+            if obj is None:
+                return False
+            return obj.__module__.startswith('napari.')
+
+        except AttributeError:
+            return False
```
Comment (jni) [inline napari/utils/events/event.py]: I found this flow hard to follow, because I don't know all of the things that `callback` can be. Perhaps some comments would help? Like, what is `obj` expected to be?

I think I would also prefer a path that unifies the two non-False returns.

**Comment by jni** [comment]
@Czaki I don't fully understand what these changes address. Could you explain the core motivation a bit better? I am also reluctant to merge during the rc period because they seem to potentially affect a lot of behaviour, and they don't address a critical bug or regression (?), so I think it can probably wait until 0.4.17?

File: napari/utils/events/event.py (hunk 3 above)
Comment (Czaki) [reply napari/utils/events/event.py]: Callback could be `Tuple[weakref, str[` representing a reference to python object and method name or function. So `obj` is derefered object which method callback should call. 

Did I should add documentation in this function, or next to the definition of `CallbackRef` and `Callback`?

**Comment by Czaki** [comment]
> @Czaki I don't fully understand what these changes address. Could you explain the core motivation a bit better? I am also reluctant to merge during the rc period because they seem to potentially affect a lot of behaviour, and they don't address a critical bug or regression (?), so I think it can probably wait until 0.4.17?

In this PR I try to ensure that all napari internal callbacks were called before any external callback. In the last weeks were at least two issues connected with the napari crash because the plugin creator callback was called before napari internal callbacks when the internal state was inconsistent.

**Comment by brisvag** [comment]
(For context, the issues were linked by @Czaki in `References`; maybe @jni you missed them?)

**Comment by Czaki** [comment]
And maybe a small addition why not simply change the default value for position to `last`. 
This PR does not change the order of execution of internal napari callbacks. It also does not change the order of user-defined callbacks. It only moves all user-defined callbacks on the end, preserving order.

File: napari/utils/events/event.py (hunk 3 above)
Comment (jni) [reply napari/utils/events/event.py]: Comments describing this are fine. For me the guideline is that documentation should be local — I should be able to understand well enough a bit of code without having to jump around the codebase.

File: napari/utils/_tests/test_notification_manager.py (hunk 2 above)
Comment (Czaki) [reply napari/utils/_tests/test_notification_manager.py]: done

File: napari/utils/events/event.py (hunk 3 above)
Comment (Czaki) [reply napari/utils/events/event.py]: @jni could you check docstring and comments? 

And I do not see good way to merge this two way of non False returns because these are two different objects. One is a function, the second one is an object. 

The only option that I see is to get the method from the object and assign it to the callback. but this introduces additional complexity for me.

**Review by sofroniewn — APPROVED** [review]
I didn't love how napari specific this was making our event file, as it might be one day nice to remove this from the repo and put our event system in pysignal or somewhere else, so with @jni we renamed the `_is_core_napari_callback` and made it now take an additional kwarg. This might make it a bit easier to de-naparify this file at some point in time in the future.

I think longer term it would also be nice to get to a point where we were connecting events for plugins rather than having the connections done on the plugin side

**Comment by Czaki** [comment]
> I didn't love how napari specific this was making our event file, as it might be one day nice to remove this from the repo and put our event system in pysignal or somewhere else, so with @jni we renamed the `_is_core_napari_callback` and made it now take an additional kwarg. This might make it a bit easier to de-naparify this file at some point in time in the future.

I personally prefer to totally split napari internal events and napari public events. 

something like:

```
self.events._inserted(data) # for internal callback
self.events.inserted(data) # for plugin callback
```

But I have not enough time to explore whole codebase and update it so i prepare semi automatic workaround.

**Comment by jni** [comment]
Thanks @Czaki! Yeah, apologies that I missed the motivation in the references either. Maybe we should put references at the top of the PR template!

**Review by jni — APPROVED** [review]
